enterprise/backend/src/metabase_enterprise/workspaces/impl.clj

```diff
@@ -0,0 +1,41 @@
+(ns metabase-enterprise.workspaces.impl
+  "Glue code connecting workspace subsystems (dependencies, isolation)."
+  (:require
+   [metabase-enterprise.workspaces.dependencies :as ws.deps]
+   [metabase-enterprise.workspaces.isolation :as ws.isolation]
+   [toucan2.core :as t2]))
+
+(defn- query-external-inputs
+  "Query for external inputs in a workspace - inputs that are not shadowed by any output.
+   Returns seq of WorkspaceInput records."
+  [workspace-id]
+  ;; TODO: maybe we should have a flag on workspace_input.access_granted ?
+  (t2/select :model/WorkspaceInput
+             :workspace_id workspace-id
+             {:where [:not [:exists {:select [1]
```

Comment:
Optimization idea: use `table_id` to join faster when we have it on both sides (short-circuit the logic inside the where clause)